File: src/is/skeletons-detector-service/stream.cpp
#include <chrono>
#include <cstdlib>
#include <vector>
#include <numeric>
#include "skeletons/utils.hpp"
#include "skeletons/detector.hpp"
#include <is/wire/core.hpp>
#include <is/msgs/utils.hpp>
#include <prometheus/exposer.h>
#include <prometheus/registry.h>

using namespace std::chrono;

int main(int argc, char** argv) {
  auto options = load_options(argc, argv);
  const std::string service_name{"SkeletonsDetector"};

  auto channel = StreamChannel(options.broker_uri());
  is::info("Connected to broker {}", options.broker_uri());
  auto subscription = is::Subscription(channel.internal_channel(), service_name + ".Detection");
  subscription.subscribe("CameraGateway.*.Frame");
  auto tracer = make_tracer(options, "SkeletonsDetector");

  SkeletonsDetector detector(options);
  auto gpu_device_id_var = std::getenv("GPU_DEVICE_ID");
  auto gpu_device_id = gpu_device_id_var != nullptr ? gpu_device_id_var : "";
  int dropped;


  prometheus::Exposer exposer("8080");  // Expose metrics at http://<ip>:8080/metrics
  auto registry = std::make_shared<prometheus::Registry>();
  exposer.RegisterCollectable(registry);
  auto& skeletons_metric =
      prometheus::BuildGauge().Name("skeletons").Register(*registry).Add({});
  skeletons_metric.Set(0.0);

  std::vector<double> buffer;
  
  steady_clock::time_point start;
  start = steady_clock::now();

  while (true) {
    const auto start_span = [&](auto& maybe_ctx, auto span_name) {
      return maybe_ctx ? tracer->StartSpan(span_name, {ChildOf(maybe_ctx->get())}) : tracer->StartSpan(span_name);
    };

    auto msg = channel.consume_last(&dropped);

    auto maybe_ctx = msg.extract_tracing(tracer);
    auto service_span = start_span(maybe_ctx, "detection_and_render");

    auto unpack_span = start_span(maybe_ctx, "unpack");
    auto pb_image = msg.unpack<is::vision::Image>();
    if (!pb_image) continue;
    unpack_span->Finish();

    auto t0 = steady_clock::now();
    auto detection_span = start_span(maybe_ctx, "detection");
    auto camera_id = get_topic_id(msg.topic());
    auto skeletons = detector.detect(pb_image.get(), camera_id);

    detection_span->SetTag("gpu_device_id", gpu_device_id);
    detection_span->SetTag("detections", skeletons.objects_size());
    detection_span->Finish();

    auto t1 = steady_clock::now();
    auto pack_sk_span = start_span(maybe_ctx, "pack_and_publish_detections");
    is::Message skeletons_msg{skeletons};
    skeletons_msg.inject_tracing(tracer, service_span->context());
    channel.publish(fmt::format("SkeletonsDetector.{}.Detection", camera_id), skeletons_msg);
    pack_sk_span->Finish();

    auto render_span = start_span(maybe_ctx, "render_pack_publish_expose");
    auto cv_image = detector.last_image();
    Image pb_rendered_image;
    render_skeletons(cv_image, skeletons, &pb_rendered_image);
    is::Message rendered_msg{pb_rendered_image};
    channel.publish(fmt::format("SkeletonsDetector.{}.Rendered", camera_id), rendered_msg);

    auto end = steady_clock::now();
    auto duration = duration_cast<seconds>(end - start);
    if ((duration.count() >= options.period()) && (buffer.size() > 0)) {
      double mean = std::accumulate(buffer.begin(), buffer.end(), 0.0) / buffer.size();
      skeletons_metric.Set(mean);
      buffer.clear();
      start = steady_clock::now();
    }
    else {
      auto num_skeletons = (double) skeletons.objects_size();
      buffer.push_back(num_skeletons);
    }

    render_span->Finish();

    service_span->Finish();
    auto t2 = steady_clock::now();

    const auto dt = [](auto& tf, auto& t0) { return duration_cast<microseconds>(tf - t0).count() / 1000.0; };
    is::info("source = {}, detections = {}, dropped_messages = {}", msg.topic(), skeletons.objects_size(), dropped);
    is::info("took_s = {{ detection = {:.2f}, service = {:.2f} }}", dt(t1, t0), dt(t2, t0));
  }
}

File: src/is_skeletons_detector/stream.py
import re
import time
import dateutil.parser as dp

from is_wire.core import Subscription, Message, Logger, Tracer, AsyncTransport
from opencensus.ext.zipkin.trace_exporter import ZipkinExporter
from is_msgs.image_pb2 import Image
from prometheus_client import start_http_server, Gauge

from .stream_channel import StreamChannel
from .skeletons import SkeletonsDetector
from .utils import load_options, get_np_image, get_pb_image, draw_skeletons


def span_duration_ms(span):
    dt = dp.parse(span.end_time) - dp.parse(span.start_time)
    return dt.total_seconds() * 1000.0


def create_exporter(service_name, uri):
    log = Logger(name="CreateExporter")
    zipkin_ok = re.match("http:\\/\\/([a-zA-Z0-9\\.]+)(:(\\d+))?", uri)
    if not zipkin_ok:
        log.critical("Invalid zipkin uri \"{}\", expected http://<hostname>:<port>", uri)
    exporter = ZipkinExporter(
        service_name=service_name,
        host_name=zipkin_ok.group(1),
        port=zipkin_ok.group(3),
        transport=AsyncTransport)
    return exporter


def main():
    service_name = 'SkeletonsDetector.Detection'
    re_topic = re.compile(r'CameraGateway.(\w+).Frame')

    op = load_options()
    sd = SkeletonsDetector(op)

    log = Logger(name=service_name)
    channel = StreamChannel(op.broker_uri)
    log.info('Connected to broker {}', op.broker_uri)

    exporter = create_exporter(service_name=service_name, uri=op.zipkin_uri)
    
    skeletons_detected = Gauge("skeletons",
                               "Skeletons detected by any camera")
    skeletons_detected.set(0.0)
    start_http_server(8080)

    buffer = list()
    initial_time = time.time()

    subscription = Subscription(channel=channel, name=service_name)
    subscription.subscribe('CameraGateway.*.Frame')

    while True:
        msg, dropped = channel.consume(return_dropped=True)

        tracer = Tracer(exporter, span_context=msg.extract_tracing())
        span = tracer.start_span(name='detection_and_render')
        detection_span = None

        with tracer.span(name='unpack'):
            im = msg.unpack(Image)
            im_np = get_np_image(im)
        with tracer.span(name='detection') as _span:
            skeletons = sd.detect(im_np)
            detection_span = _span
        with tracer.span(name='pack_and_publish_detections'):
            sks_msg = Message()
            sks_msg.topic = re_topic.sub(r'SkeletonsDetector.\1.Detection', msg.topic)
            sks_msg.inject_tracing(span)
            sks_msg.pack(skeletons)
            channel.publish(sks_msg)
        with tracer.span(name='render_pack_publish_expose'):
            im_rendered = draw_skeletons(im_np, skeletons)
            rendered_msg = Message()
            rendered_msg.topic = re_topic.sub(r'SkeletonsDetector.\1.Rendered', msg.topic)
            rendered_msg.pack(get_pb_image(im_rendered))
            channel.publish(rendered_msg)
            buffer.append(len(skeletons.objects))
            if (time.time() - initial_time) > op.period and len(buffer) > 0:
                log.info('metric = {:5.2f}',sum(buffer) / len(buffer))
                skeletons_detected.set(sum(buffer) / len(buffer))
                buffer = []
                initial_time = time.time()

        span.add_attribute('Detections', len(skeletons.objects))
        tracer.end_span()
        log.info('detections = {:2d}, dropped_messages = {:2d}', len(skeletons.objects), dropped)
        log.info('took_ms = {{ detection: {:5.2f}, service: {:5.2f}}}',
                 span_duration_ms(detection_span), span_duration_ms(span))


if __name__ == "__main__":
    main()